src/extract/pathbuf.rs

```diff
@@ -1,116 +1,101 @@
-use extract::{Context, Error, Extract, ExtractFuture};
-use futures::Poll;
+use extract::{Context, Error, Extract, Immediate};
+use std::borrow::Cow;
+use std::ffi::OsStr;
 use std::path::{self, Path, PathBuf};
+use url::percent_encoding;
 use util::buf_stream::BufStream;
 
-impl<B: BufStream> Extract<B> for PathBuf {
-    type Future = ExtractPathBuf;
-
-    fn extract(ctx: &Context) -> Self::Future {
-        use codegen::Source::*;
-
-        match ctx.callsite().source() {
-            Capture(idx) => {
-                let path = ctx.request().uri().path();
-                let value = ctx.captures().get(*idx, path).into();
-                ExtractPathBuf(value)
-            }
-            _ => unimplemented!("A PathBuf can only be extracted from a path capture for now"),
-        }
-    }
+#[cfg(not(any(target_os = "windows", target_arch = "wasm32")))]
+fn osstr_from_bytes(bytes: &[u8]) -> Result<&OsStr, Error> {
+    use std::os::unix::ffi::OsStrExt;
+    Ok(OsStr::from_bytes(bytes))
 }
 
-/// Extracts a `PathBuf` value from an HTTP request.
-#[derive(Debug)]
-pub struct ExtractPathBuf(String);
+#[cfg(any(target_os = "windows", target_arch = "wasm32"))]
```

Comment:
The two implementations feel imbalanced. It seems like if we want to have the ability to provide special values for *nix, then it should be present for Windows as well. Conversely, if Windows only supports UTF-8, then it feels like so should *nix.

Serde [has some code](https://github.com/serde-rs/serde/blob/3da0deaa5073f7f9367e95f347bd80ac01a5dd6a/serde/src/de/impls.rs#L1551-L1581) where they parse in either a `Vec<u8>` or `Vec<u16>` and then call the appropriate method ([unix](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStringExt.html#tymethod.from_vec) / [windows](https://doc.rust-lang.org/std/os/windows/ffi/trait.OsStringExt.html#tymethod.from_wide)).

Reply:
I'm not aware of a standard way to percent-encode UTF-16 or UCS-2 URLs.